File: test_e2e.py
#!/usr/bin/env python3
"""
End-to-end pipeline validation against reference DaliLite output.

For each reference pair, runs the full pipeline (both wolf and parsi paths,
both directions) and compares DCCP output against reference.
"""

import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from pipeline import compare_pair
from validation.parsers import parse_dccp

DAT_DIR = Path('validation/fixtures/structures')
E2E_DIR = Path('validation/fixtures/e2e')

PAIRS = [
    ('101mA', '1a00A'),
    ('101mA', '1binA'),
    ('1a87A', '1allA'),
]

# parse_dccp stores zmax in the .rmsd field and rmsd in the .zscore field
# due to positional parsing. Our DCCPResult has the fields named correctly.
SCORE_TOL = 2.0


def compare_dccp_blocks(cand_results, ref_blocks, cd1, cd2, label):
    """Compare candidate DCCPResults against reference DCCPBlocks.

    Groups by direction (cd1->cd2 vs cd2->cd1) and compares scores.

    Returns:
        (matched, total, details)
    """
    details = []
    matched = 0
    total = 0

    for direction_cd1, direction_cd2, dir_label in [
        (cd1, cd2, f'{cd1}->{cd2}'),
        (cd2, cd1, f'{cd2}->{cd1}'),
    ]:
        ref_dir = [b for b in ref_blocks
                   if b.cd1 == direction_cd1 and b.cd2 == direction_cd2]
        cand_dir = [r for r in cand_results
                    if r.cd1 == direction_cd1 and r.cd2 == direction_cd2]

        if not ref_dir and not cand_dir:
            continue

        total += max(len(ref_dir), len(cand_dir))

        if not ref_dir:
            details.append(f"    {dir_label}: {len(cand_dir)} EXTRA results (ref has 0)")
            continue
        if not cand_dir:
            details.append(f"    {dir_label}: MISSING (ref has {len(ref_dir)})")
            continue

        # Compare count
        if len(cand_dir) != len(ref_dir):
            details.append(f"    {dir_label}: count mismatch cand={len(cand_dir)} "
                           f"ref={len(ref_dir)}")

        # Compare scores: ref.rmsd is actually zmax due to parser field order
        for i in range(min(len(cand_dir), len(ref_dir))):
            r = cand_dir[i]
            ref = ref_dir[i]
            score_diff = abs(r.score - ref.score)
            # ref.rmsd holds zmax, ref.zscore holds rmsd
            z_diff = abs(r.zmax - ref.rmsd)

            ok_score = score_diff <= SCORE_TOL
            ok_z = z_diff <= 0.5

            if ok_score:
                matched += 1
                details.append(
                    f"    {dir_label}[{i}]: MATCH score={r.score:.1f} "
                    f"(ref={ref.score:.1f} diff={score_diff:.1f}) "
                    f"z={r.zmax:.1f} (ref={ref.rmsd:.1f} diff={z_diff:.1f})")
            else:
                details.append(
                    f"    {dir_label}[{i}]: DIFF  score={r.score:.1f} "
                    f"(ref={ref.score:.1f} diff={score_diff:.1f}) "
                    f"z={r.zmax:.1f} (ref={ref.rmsd:.1f} diff={z_diff:.1f})")

    return matched, total, details


def test_pair(cd1, cd2):
    """Run full pipeline on one pair and compare against reference."""
    pair_dir = E2E_DIR / f'{cd1}_vs_{cd2}'
    ref_path = pair_dir / f'{cd1}.dccp'

    ref_blocks = parse_dccp(str(ref_path))
    # Filter self-comparisons from reference (shouldn't be any, but just in case)
    ref_blocks = [b for b in ref_blocks if b.cd1 != b.cd2]

    print(f"\n  --- {cd1} vs {cd2} ---")
    print(f"  Reference: {len(ref_blocks)} DCCP blocks")

    t0 = time.time()
    results = compare_pair(cd1, cd2, str(DAT_DIR))
    elapsed = time.time() - t0
    print(f"  Pipeline:  {len(results)} DCCP results ({elapsed:.0f}s)")

    matched, total, details = compare_dccp_blocks(
        results, ref_blocks, cd1, cd2, f'{cd1}_vs_{cd2}')

    for d in details:
        print(d)

    return matched, total


def main():
    print("=" * 60)
    print("End-to-End Pipeline Validation")
    print("=" * 60)

    total_matched = 0
    total_count = 0

    for cd1, cd2 in PAIRS:
        matched, total = test_pair(cd1, cd2)
        total_matched += matched
        total_count += total

    print(f"\n{'=' * 60}")
    print(f"TOTAL: {total_matched}/{total_count} direction-entries matched "
          f"(score tolerance={SCORE_TOL})")
    print(f"{'=' * 60}")

    return total_matched == total_count


if __name__ == '__main__':
    ok = main()
    sys.exit(0 if ok else 1)

File: dali_rust/dali-core/tests/test_e2e.rs
use std::fs;

use dali_core::pipeline::compare_pair;
use dali_core::ProteinStore;
use dali_core::DccpEntry;

const FIXTURE_DIR: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../..",
    "/validation/fixtures/structures"
);
const E2E_DIR: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../..",
    "/validation/fixtures/e2e"
);

/// Parsed reference DCCP entry from ground truth file.
#[derive(Debug)]
struct DccpRef {
    score: f64,
    zmax: f64,
    cd1: String,
    cd2: String,
}

/// Parse a reference DCCP file into entries (skipping alignment lines).
fn parse_dccp_ref(path: &str) -> Vec<DccpRef> {
    let content = fs::read_to_string(path).unwrap();
    let mut entries = Vec::new();

    for line in content.lines() {
        let trimmed = line.trim();
        if !trimmed.starts_with("DCCP") {
            continue;
        }
        let parts: Vec<&str> = trimmed.split_whitespace().collect();
        if parts.len() < 10 {
            continue;
        }

        // Format: DCCP 1 score rmsd 0 zmax 0 nblock cd1 cd2
        let score: f64 = match parts[2].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let zmax: f64 = match parts[5].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let cd1 = parts[8].to_string();
        let cd2 = parts[9].to_string();

        entries.push(DccpRef { score, zmax, cd1, cd2 });
    }

    entries
}

/// Compare candidate DccpEntry results against reference, grouped by direction.
///
/// Returns (matched, total, mismatch details).
fn validate_pair(
    candidates: &[DccpEntry],
    refs: &[DccpRef],
    cd1: &str,
    cd2: &str,
    score_tol: f64,
) -> (usize, usize, Vec<String>) {
    let mut matched = 0;
    let mut total = 0;
    let mut details = Vec::new();

    for (dir_cd1, dir_cd2) in [(cd1, cd2), (cd2, cd1)] {
        let ref_dir: Vec<&DccpRef> = refs.iter()
            .filter(|r| r.cd1 == dir_cd1 && r.cd2 == dir_cd2)
            .collect();
        let cand_dir: Vec<&DccpEntry> = candidates.iter()
            .filter(|r| r.cd1 == dir_cd1 && r.cd2 == dir_cd2)
            .collect();

        if ref_dir.is_empty() && cand_dir.is_empty() {
            continue;
        }

        total += ref_dir.len().max(cand_dir.len());

        if ref_dir.is_empty() {
            details.push(format!(
                "  {}→{}: {} EXTRA results (ref has 0)",
                dir_cd1, dir_cd2, cand_dir.len()
            ));
            continue;
        }
        if cand_dir.is_empty() {
            details.push(format!(
                "  {}→{}: MISSING (ref has {})",
                dir_cd1, dir_cd2, ref_dir.len()
            ));
            continue;
        }

        for i in 0..ref_dir.len().min(cand_dir.len()) {
            let r = &ref_dir[i];
            let c = &cand_dir[i];
            let score_diff = (r.score - c.score).abs();

            if score_diff <= score_tol {
                matched += 1;
                details.push(format!(
                    "  {}→{}[{}]: MATCH score={:.1} (ref={:.1} diff={:.1})",
                    dir_cd1, dir_cd2, i, c.score, r.score, score_diff
                ));
            } else {
                details.push(format!(
                    "  {}→{}[{}]: DIFF  score={:.1} (ref={:.1} diff={:.1})",
                    dir_cd1, dir_cd2, i, c.score, r.score, score_diff
                ));
            }
        }
    }

    (matched, total, details)
}

#[test]
fn test_e2e_pipeline() {
    let store = ProteinStore::new(FIXTURE_DIR);
    let score_tol = 2.0;

    let pairs = [
        ("101mA", "1a00A"),
        ("101mA", "1binA"),
        ("1a87A", "1allA"),
    ];

    let mut total_matched = 0;
    let mut total_count = 0;

    for (cd1, cd2) in &pairs {
        let ref_path = format!("{}/{}_vs_{}/{}.dccp", E2E_DIR, cd1, cd2, cd1);
        let refs = parse_dccp_ref(&ref_path);
        // Filter self-comparisons from reference
        let refs: Vec<DccpRef> = refs.into_iter()
            .filter(|r| r.cd1 != r.cd2)
            .collect();

        eprintln!("\n  --- {} vs {} ---", cd1, cd2);
        eprintln!("  Reference: {} DCCP entries", refs.len());

        let results = compare_pair(cd1, cd2, &store);
        eprintln!("  Pipeline:  {} DCCP results", results.len());

        let (matched, count, details) = validate_pair(
            &results, &refs, cd1, cd2, score_tol,
        );

        for d in &details {
            eprintln!("{}", d);
        }

        total_matched += matched;
        total_count += count;
    }

    eprintln!("\nE2E TOTAL: {}/{} matched (score_tol={})", total_matched, total_count, score_tol);

    assert!(
        total_matched >= 8,
        "E2E: only {}/{} matched, expected >=8",
        total_matched, total_count,
    );
}
